src/mito2/src/memtable/time_series.rs

```diff
@@ -0,0 +1,840 @@
+                v.insert((bucket_idx, series_index));
+                s
+            }
+            // safety: series must exist at given index.
+            Entry::Occupied(v) => self.get_bucket_by_index(v.get()).unwrap(),
+        }
+    }
+
+    /// Iterates all series in [SeriesSet].
+    fn iter_series(&self) -> Iter {
+        Iter {
+            metadata: self.region_metadata.clone(),
+            indices: self.indices.clone(),
+            buckets: self.buckets.clone(),
+            last_key: None,
+        }
+    }
+
+    /// Hash primary key to find the index of bucket.
+    fn hash_primary_key(&self, primary_key: &[u8]) -> u8 {
+        let x: u32 = primary_key.iter().map(|v| *v as u32).sum();
+        (x % self.bucket_num as u32) as u8
+    }
+}
+
+struct Iter {
+    metadata: RegionMetadataRef,
+    indices: Arc<RwLock<BTreeMap<Vec<u8>, SeriesIndex>>>,
+    buckets: Arc<Vec<Bucket>>,
+    last_key: Option<Vec<u8>>,
+}
+
+impl Iterator for Iter {
+    type Item = Result<Batch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let map = self.indices.read().unwrap();
+        let mut range = match &self.last_key {
+            None => map.range::<Vec<u8>, _>(..),
+            Some(last_key) => {
+                map.range::<Vec<u8>, _>((Bound::Excluded(last_key), Bound::Unbounded))
+            }
+        };
+
+        if let Some((primary_key, bucket)) = range.next() {
+            self.last_key = Some(primary_key.clone());
+            let Some(b) = self.buckets.get(bucket.0 as usize) else {
+                panic!("Cannot find series at {:?}", bucket);
+            };
+            let Some(series) = b.get_series(bucket.1) else {
+                panic!("Cannot find series at {:?}", bucket);
+            };
+            let values = series.write().unwrap().compact();
+            Some(values.and_then(|v| v.to_batch(primary_key, &self.metadata)))
+        } else {
+            None
+        }
+    }
+}
+
+/// Bucket holds a set of [Series] which alleviate lock contention between series.
+struct Bucket {
+    region_metadata: RegionMetadataRef,
+    series: RwLock<Vec<Arc<RwLock<Series>>>>,
+}
+
+impl Bucket {
+    fn new(region_metadata: RegionMetadataRef) -> Self {
+        Self {
+            region_metadata,
+            series: Default::default(),
+        }
+    }
+
+    /// Returns the series at given index.
+    /// Returns None if series not found.
+    #[inline]
+    fn get_series(&self, idx: usize) -> Option<Arc<RwLock<Series>>> {
+        self.series.read().unwrap().get(idx).cloned()
+    }
+
+    /// Adds series to bucket and returns the index inside the bucket.
+    #[inline]
+    fn add_series(&self, s: Arc<RwLock<Series>>) -> usize {
+        let mut series = self.series.write().unwrap();
+        let idx = series.len();
+        series.push(s);
+        idx
+    }
+}
+
+/// A `Series` holds a list of field values of some given primary key.
+struct Series {
+    initial_builder_capacity: usize,
```

Comment:
Could we define this as a constant or pass it at each function call? Storing this at each series is a little wasteful. Same as the RegionMetadataRef.